sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1653,6 +1978,106 @@ submitted in its entirety for execution via
   ...
 ----
 
+=== Dynamic Parameter Update
+
+Example showing a graph with a single kernel node that is created using a kernel
+bundle with `handler::set_args()` and having its node arguments updated.
+
+[source,c++]
+----
+...
+
+using namespace sycl;
+namespace sycl_ext = sycl::ext::oneapi::experimental;
+
+queue myQueue;
+auto myContext = myQueue.get_context();
+auto myDevice = myQueue.get_device();
+
+// USM allocations for kernel input/output
+const size_t n = 1024;
+int *ptrX = malloc_shared<int>(n, myQueue);
+int *ptrY = malloc_device<int>(n, myQueue);
+
+int *ptrZ = malloc_shared<int>(n, myQueue);
+int *ptrQ = malloc_device<int>(n, myQueue);
+
+// Kernel loaded from kernel bundle
+const std::vector<kernel_id> builtinKernelIds =
+      myDevice.get_info<info::device::built_in_kernel_ids>();
+kernel_bundle<bundle_state::executable> myBundle =
+      get_kernel_bundle(myContext, { myDevice }, builtinKernelIds);
+kernel builtinKernel = myBundle.get_kernel(builtinKernelIds[0]);
+
+// Graph containing a two kernels node
+sycl_ext::command_graph myGraph(myContext, myDevice);
+
+int myScalar = 42;
+// Create graph dynamic parameters
+dynamic_parameter dynParamInput(myGraph, ptrX);
+dynamic_parameter dynParamScalar(myGraph, myScalar);
+
+// First node uses ptrX as an input & output parameter, with operand
+// mySclar as another argument.
+node nodeA = myGraph.add([&](handler& cgh) {
+    cgh.set_args(ptrX, ptrY, myScalar);
+    // Register nodeA dynamic parameters
+    dynParamInput.register_with_node(cgh, 0);  // Argument index 0 is ptrX
+    dynParamScalar.register_with_node(cgh, 2);  // Argument index 2 is myScalar
+    cgh.parallel_for(range {n}, builtinKernel);
+});
+
+
+// Create an executable graph with the updatable property.
+auto execGraph = myGraph.finalize({sycl_ext::property::graph::updatable});
+
+// Execute graph, then update without needing to wait for it to complete
+myQueue.ext_oneapi_graph(execGraph);
+
+// Change ptrX argument to node A to ptrZ
+dynParamInput.update(ptrZ);
+
+// Change myScalar argument to node A to newScalar
+int newScalar = 12;
+dynParamScalar.update(newScalar);
+
+// Update nodeA in the executable graph with the new parameters
+execGraph.update(nodeA);
+// Execute graph again
+myQueue.ext_oneapi_graph(execGraph);
+myQueue.wait();
+
+sycl::free(ptrX, myQueue);
+sycl::free(ptrY, myQueue);
+sycl::free(ptrZ, myQueue);
+sycl::free(ptrQ, myQueue);
+
+----
+
+Example snippet showing how to use accessors with `dynamic_parameter` update:
+[source,c++]
+----
+sycl::buffer bufferA{...};
+sycl::buffer bufferB{...};
+
+// Create graph dynamic parameter using a placeholder accessor, since the
+// sycl::handler is not available here outside of the command-group scope.
+dynamic_parameter dynParamAccessor(myGraph, bufferA.get_access());
+
+node nodeA = myGraph.add([&](handler& cgh) {
+    // Create a non-placeholder accessor for the kernel
+    auto accA = bufferA.get_access(cgh);
+    cgh.set_args(accA);
```

Comment:
This API seems a bit unsafe because there is no way to ensure that the parameter to `cgh.set_args` matches the `dynamicAccessor`.  For example, the user might mistakenly type `bufferB.get_access(cgh)`, and that would not be caught by the compiler.  Presumably, things would just fail in a weird way later when the parameter was updated.

This got me thinking that maybe the `dynamic_parameter` API could be improved overall, not just for accessors.  Could we remove the `register_with_node` function entirely and instead pass the `dynamic_parameter` directly to `set_args`?  For example:

```
node nodeA = myGraph.add([&](handler& cgh) {
    cgh.require(dynParamAccessor);  // Only necessary when dynamic_parameter is a placeholder accessor
    cgh.set_args(dynParamAccessor);
    cgh.parallel_for(range {n}, builtinKernel);
});
```

The call to `cgh.set_args` can encompass the functionality of `register_with_node` by having a special overload when the argument is `dynamic_parameter`.  We know the argument index implicitly because of the position in the argument list.

I think there is no need to change the formal parameters to the kernel function.  Even though, the user passes a `dynamic_parameter` to `set_args`, the kernel function parameter can still be the underlying type (in this case `accessor`).

Reply:
I agree this seems like an improvement for dynamic parameter usage and helps with the case you mentioned as well. As a knock-on effect this would make the value passed to the `dynamic_parameter` constructor needed for more than just CTAD as it would now be the initial value of the parameter, which is probably also a positive change. Might require some clarifying wording on `dynamic_parameter::update()` also about the internal value state.

Implementation might be a little tricky with this so I will prototype the implementation first before committing to the spec change.

Reply:
I've been prototyping this change and in general I am fine with it, however I think we run into a few problems due to the lack of an implementation for free function kernels (the major intended use case). It is possible to test it with regular kernels, but this is undefined behaviour according to the spec, so not really a viable option. Interop kernels might be possible but have a very high cost for implementing in the tests in a backend agnostic way. So until the free function extension is implemented I don't think we can remove the existing mechanism, and thus our only sensible way of testing right now.

I am not keen to delay this work for a potentially significant amount of time waiting for an implementation of that extension. So the options as I see them are:

- Keep both mechanisms in the spec, with a note (similar to the base spec) that using `set_arg` with regular kernels is undefined behaviour. That way we can keep testing with the old API (the code paths are largely the same outside of the brief interactions with the handler) and just remove it and update the tests when the extension is available.
- Keep only the old API for now, and make this change to the spec (and implementation) at a later date (or include the spec changes in the future work section). This makes the spec and implementation more consistent and removes potential confusion, but leaves this as an open TODO which is not that desirable.

Reply:
> I've been prototyping this change and in general I am fine with it, however I think we run into a few problems due to the lack of an implementation for free function kernels (the major intended use case). It is possible to test it with regular kernels, but this is undefined behaviour according to the spec, so not really a viable option.

I'm glad to hear that the prototype was going well.  I think we should try to find a way to push forward with our preferred API, though, rather than reverting to an API that we don't like.  You say that it is possible to test the preferred API with regular kernels, so I wonder if we can use this as a temporary solution until free function kernels are implemented.

Your concern about testing with regular kernels is that the behavior is undefined according to the spec.  However, I think this is also the case for the existing proposed API for `dynamic_parameter`.  The existing API for `dynamic_parameter` requires the user to specify the argument index, but there is no such thing for regular kernels.  The argument order of variables captured by a lambda is not specified by C++, so calling `dynamic_parameter::register_with_node(int argIndex, ...)` is also undefined behavior according to the spec.

I think we decided that our initial design for `dynamic_parameter` would be limited to only work for free function kernels.  This is our only important use case in the short term, so we wanted to concentrate on that first.  Once we had that working, we planned to revisit the design and see how it could be extended for regular kernels.  Given that priority, I think it makes sense to pursue an API that makes the most sense for free function kernels.

Reply:
> Your concern about testing with regular kernels is that the behavior is undefined according to the spec. However, I think this is also the case for the existing proposed API for `dynamic_parameter`. The existing API for `dynamic_parameter` requires the user to specify the argument index, but there is no such thing for regular kernels. The argument order of variables captured by a lambda is not specified by C++, so calling `dynamic_parameter::register_with_node(int argIndex, ...)` is also undefined behavior according to the spec.

My main concern is really the fragility of the code path in the new implementation (even though both are undefined/unspecified as you say), since it is much more dependent on the details of the handler's implementation. After some consideration though, if the tests are running on CI then even if it is more likely to break those breaks should still be caught before being merged, so it should be safe enough.

I'll move forward with the API for now then as suggested. I'll push an update to the spec soon once I've ironed out some kinks in the implementation.

Reply:
I've updated the spec to reflect these changes.